django_netjsonconfig/tests/test_template.py

```diff
@@ -263,3 +263,23 @@ def test_duplicate_files_in_template(self):
             self.assertIn('Invalid configuration triggered by "#/files"', str(e))
         else:
             self.fail('ValidationError not raised!')
+
+    def test_variable_substition(self):
```

Comment:
Not sure, If I understand the purpose of this test.  Is it intended to be this way?

Reply:
It is there to ensure validation doesn't fail when saving a config with variables. This test ensures that the variables in config are replaced during config validation.  How did you think it was intended to be?

Reply:
Okay, correct me if I am wrong, the purpose of this is to ensure that the context variable is changed with it's value in the output, right? i.e`obj.config['dns_server']` should have `4.4.4.4` in it? 
Secondly,
```
obj = self.template_model.objects.get(name='test1')
self.assertEqual(obj.name, 'test1')
```
Here, assert Equal can never fail here, either obj.name is `test1` or `self.template_model.objects.get(name='test1')` returns an error and test errors out instead of failing! :smile:

Reply:
no @atb00ker . the value of `obj.config['dns_server']` is not actually changed. the `default_values` is just use to trick the `schema` to think that the config is correct and to replace the `context` when the `template` is about to be applied to a device. it doesn't replace the `{{dns}}` in the database. `{{dns}}` is still been saved to the database. so in effect we are testing this [line](https://github.com/openwisp/django-netjsonconfig/pull/178/commits/c1554ac4a159dfe99d9b703252bdbb265e465c52#diff-dd64f186cf77b4ab23c7cdcd1cc61cd6R150). Also, if `{{dns}}` is not replaced by `default_values` during validation, the test will fail because the schema rejected the `config` as it was not valid and the object was not created. I don't know if i have explained it well :smile:

Reply:
Ah, that explains it, I misunderstood the purpose of the test, thanks! :smile: